core/src/main/scala/cats/syntax/foldable.scala

```diff
@@ -34,6 +34,12 @@ final class FoldableOps[F[_], A](val fa: F[A]) extends AnyVal {
   def foldr[B](b: Eval[B])(f: (A, Eval[B]) => Eval[B])(implicit F: Foldable[F]): Eval[B] =
     F.foldRight(fa, b)(f)
 
+  /**
+   * given a Monoid evidence for `A`, it returns None if the foldable is empty or combines all the `A`s if it's not
+   */
+  def combineAllOption(implicit ev: Monoid[A], F: Foldable[F]): Option[A] =
+    if (F.isEmpty(fa)) None else Some(F.combineAll(fa))
```

Comment:
This should probably be using `reduceLeftOption ` and `Semigroup.combine`

Reply:
If we can't use `Monoid.combineAll` or `Semigroup.combineAllOption` we are missing the optimization. This is for cases when we can use internal mutability to aggregate many fast, which is pretty critical in some applications (e.g. spark/scalding).

If we just call out to `combine` we are destroying optimizations.

Reply:
Yeah, it's tricky here,  we don't want to impose `Monoid`and I wasn't aware there is optimization for `Semigroup.combineAllOption`  (actually I still can't find it), but even it does, we are in a syntax extension, so unlike `fold`, we can't optimize `combineAllOption` in `Foldable` instances. 

So to retain the optimization we probably need the keep the `Monoid` requirement which is rather unfortunate, we'd better document the rationale. Another option is to wait until Cats 2.0 with which we'll be able to add a combineAllOption to `Foldable`

In the meantime @johnynek would you mind point me to the where the `Semigroup.combineAllOption`'s optimization is? we should probably document it somewhere.

Reply:
so, here is the method:

https://github.com/typelevel/cats/blob/master/kernel/src/main/scala/cats/kernel/Semigroup.scala#L39

note any semigroup can override that (as we do in other typeclasses for things like map2, product, etc.. in Applicative).

In algebird, which uses cats.kernel, we do this very frequently.

e.g.:
https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/HyperLogLog.scala#L582

https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala#L141

there are more. The point is the person implementing the Semigroup knows if there is a more efficient way to combine many of them, and should control that. Other typeclasses ideally should delegate to that to not undo the optimization.

Reply:
It's tough in this case since here you are trading off the potential added cost of materializing a list with the potential benefit of reducing an entire group together.

Many Foldables could actually create an `Iterator` to pass into `combineAllOption` on Semigroup, but since we don't have a method like that we can only make the List.

We could imagine:

```scala
def onIterator[B](fa: F[A])(itFn: Iterator[A] => B): B =
  itFn(toList(fa).iterator)
```
on Foldable in cats 2.0. In this world, if you can make an Iterator for the caller to consume. Alternatively, we could add `Fold[A, B]` to cats: 
https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/Fold.scala#L40

so we could have `def foldWith[A, B](fa: F[A])(fold: Fold[A, B]): B` on `Foldable`. Since folds ca potentially have internal mutable state, they can do much the same optimization of combineAllOption, although, not in parallel (since you can't split over trees).

Reply:
I cannot say if we could add `Fold` so I would leave the decision to others. About the other option, how could I use the `onIterator` if it were available ? Still with the semigroup `combineAllOption` or that would be alternative to depending on overriding that in Semigroup ? Because if `onIterator` wouldn't need to be overridden, could it be another syntax ?  Thanks for the explanation.

Reply:
```scala
def combineAllOption(implicit ev: Semigroup[A], F: Foldable[F]): Option[A] =
   if (F.isEmpty(fa)) None else F.onIterator(ev.combineAllOption(_))
```

Reply:
I see. Right. Still aiming at mitigating the possible cost of building a list, it could be delayed up to when the iterator is traversing it. Something like
```scala
def combineAllOption(implicit ev: Semigroup[A], F: Foldable[F]): Option[A] =
  if (F.isEmpty(fa)) None else ev.combineAllOption(F.toList(fa).iterator)

def toIterator(implicit F: Foldable[F]): Iterator[A] =
  F.foldRight[A, Stream[A]](fa, Eval.later(Stream.empty)) {
    (a, eb) => eb map (Stream.cons(a, _))
  }.value.iterator
```
I'm not sure actually it be a step forward and is getting late. I will keep looking.